File: packages/wistro/dev/python/state_adapter.py
import json
from typing import Any

class StateAdapter:
    def __init__(self, trace_id: str, state_config: Any):
        self.trace_id = trace_id
        self.store = {}
        self.prefix = 'wistro:state:'
        self.ttl = getattr(state_config, 'ttl', None)

    def _make_key(self, key: str) -> str:
        return f"{self.prefix}{self.trace_id}:{key}"

    async def get(self, key: str) -> Any:
        full_key = self._make_key(key)
        value = self.store.get(full_key)
        return json.loads(value) if value else None

    async def set(self, key: str, value: Any) -> None:
        full_key = self._make_key(key)
        self.store[full_key] = json.dumps(value)

    async def delete(self, key: str) -> None:
        full_key = self._make_key(key)
        if full_key in self.store:
            del self.store[full_key]

    async def clear(self) -> None:
        keys_to_delete = [k for k in self.store.keys() if k.startswith(self._make_key(''))]
        for key in keys_to_delete:
            del self.store[key]

    async def cleanup(self) -> None:
        self.store.clear()

File: packages/wistro/dev/ruby/state_adapter.rb
require 'json'

class StateAdapter
  attr_reader :trace_id, :store, :prefix, :ttl

  def initialize(trace_id, state_config)
    @trace_id = trace_id
    @store = {}
    @prefix = 'wistro:state:'
    @ttl = state_config.respond_to?(:ttl) ? state_config.ttl : nil
  end

  # Generates a namespaced key
  def make_key(key)
    "#{@prefix}#{@trace_id}:#{key}"
  end

  # Retrieves a value for a given key
  def get(key)
    full_key = make_key(key)
    value = @store[full_key]
    value ? JSON.parse(value) : nil
  end

  # Stores a value for a given key
  def set(key, value)
    full_key = make_key(key)
    @store[full_key] = JSON.dump(value)
  end

  # Deletes a value for a given key
  def delete(key)
    full_key = make_key(key)
    @store.delete(full_key)
  end

  # Clears all keys for the current trace ID
  def clear
    keys_to_delete = @store.keys.select { |k| k.start_with?(make_key('')) }
    keys_to_delete.each { |key| @store.delete(key) }
  end

  # Clears the entire store
  def cleanup
    @store.clear
  end
end
